src/main/java/com/google/googleidentity/oauth2/validator/AuthorizationEndpointRequestValidator.java

```diff
@@ -0,0 +1,132 @@
+
+
+    public static void validateRedirectUri(
+            HttpServletRequest request,
+            ClientDetailsService clientDetailsService) throws OAuth2Exception {
+        String clientID = request.getParameter(OAuth2ParameterNames.CLIENT_ID);
+
+        if (Strings.isNullOrEmpty(clientID)) {
+            throw new InvalidRequestException(InvalidRequestException.ErrorCode.NO_CLIENT_ID);
+        }
+
+        Optional<ClientDetails> client = clientDetailsService.getClientByID(clientID);
+
+        if(!client.isPresent()){
+            throw new InvalidRequestException(
+                    InvalidRequestException.ErrorCode.NONEXISTENT_CLIENT_ID);
+        }
+
+        String redirectUri = request.getParameter(OAuth2ParameterNames.REDIRECT_URI);
+
+        if(Strings.isNullOrEmpty(redirectUri)){
+            throw new InvalidRequestException(InvalidRequestException.ErrorCode.NO_REDIRECT_URI);
+        }
+
+        try {
+            redirectUri = URLDecoder.decode(redirectUri, "utf-8");
+        } catch (UnsupportedEncodingException e) {
+            log.log(Level.INFO, "Uri decode failed", e);
+        }
+
+        if (!OAuth2Utils.matchUri(client.get().getRedirectUrisList(), redirectUri)) {
+            throw new InvalidRequestException(
+                    InvalidRequestException.ErrorCode.REDIRECT_URI_MISMATCH);
+        }
+    }
+
+    public static void validateGET(
+            HttpServletRequest request,
+            ClientDetailsService clientDetailsService) throws OAuth2Exception {
+
+        String clientID = request.getParameter(OAuth2ParameterNames.CLIENT_ID);
+
+        ClientDetails client = clientDetailsService.getClientByID(clientID).get();
+
+        String responseType = request.getParameter(OAuth2ParameterNames.RESPONSE_TYPE);
+
+        if(Strings.isNullOrEmpty(responseType)){
+            throw new InvalidRequestException(InvalidRequestException.ErrorCode.NO_RESPONSE_TYPE);
+        } else if (!responseType.equals("token") && !responseType.equals("code")) {
+            throw new UnsupportedResponseTypeException();
+        }
+
+        String grantType = responseType.equals("token") ? "implicit" : "authorization_code";
+
+        if(!client.getGrantTypesList().contains(grantType)){
+            throw new UnauthorizedClientException();
+        }
+
+        if (!Strings.isNullOrEmpty(request.getParameter(OAuth2ParameterNames.SCOPE))
+                && client.getIsScoped()) {
+            Set<String> scope =
+                    OAuth2Utils.parseScope(request.getParameter(OAuth2ParameterNames.SCOPE));
+            if (!client.getScopesList().containsAll(scope)) {
+                throw new InvalidScopeException();
+            }
+        }
+    }
+
+    public static void validatePOST(HttpServletRequest request) throws OAuth2Exception {
+        String userConsent = request.getParameter("user_approve");
+        String userDeny = request.getParameter("user_deny");
```

Comment:
Not sure how these two parameters are different, but can we use one parameter as boolean?